Add per player mutes

PunkPun opened the pull request:
Closes #20073 
Closes #7448 
Closes #5775 
Closes #12421

PunkPun commented:
svg's for icons can be found in https://github.com/OpenRA/ArtSrc/pull/16

<details closed>
<summary>How it looks ingame</summary>
<br>
Imagine AI's here are players
<br><br>
<img width="1087" alt="ok" src="https://media.discordapp.net/attachments/388282819371204608/993515989218099230/Screenshot_2022-07-04_at_16.57.38.png?width=1528&height=898">
<img width="1087" alt="not ok" src="https://media.discordapp.net/attachments/388282819371204608/993519391020949625/Screenshot_2022-07-04_at_17.11.16.png?width=1528&height=1122">
<img width="1087" alt="ok" src="https://media.discordapp.net/attachments/388282819371204608/993551898663723059/Screenshot_2022-07-04_at_19.20.24.png?width=1528&height=936">
<img width="1087" alt="not ok" src="https://media.discordapp.net/attachments/388282819371204608/993552641944715355/Screenshot_2022-07-04_at_19.23.23.png?width=1528&height=926">
</details>

dragunoff reviewed (commented):
Not a comprehensive review or anything, just a couple of remarks:
![image](https://user-images.githubusercontent.com/1355810/177958019-ad61a8e6-3be9-4f29-9e27-46ab4cb9edc7.png)
The button appears for the local player when spectating.

And regarding the tooltip - I think it will be better to change the text depending on the state ("Mute this player" / "Unmute this player").

PunkPun commented:
> The button appears for the local player when spectating.

Fixed. I had fixed this issue before but only for the player mute checkbox and forgot to duplicate the changes to spectator

> And regarding the tooltip - I think it will be better to change the text depending on the state ("Mute this player" / "Unmute this player").

I implemented it. Perhaps I should make tooltip toggling part of the checkbox refactor so doesn't need to be hardcoded

teinarss commented on OpenRA.Game/Network/OrderManager.cs:
```diff
@@ -55,6 +55,8 @@ public sealed class OrderManager : IDisposable
 
 		public IReadOnlyList<TextNotification> NotificationsCache => notificationsCache;
 
+		public readonly Dictionary<int, bool> MutedPlayerLib = new Dictionary<int, bool>();
```
Two things, the naming, would MutedPlayers be a better name? Also this doesn't belong on the ordermanager.

PunkPun replied:
I agree, but I'm not sure where it should belong. Order manager just fits with the usecase. I also wanted to tackle the dependancies first before going into this pr

abcdefg30 replied:
Can this live in `TextNotificationManager`?

PunkPun replied:
> Can this live in `TextNotificationManager`?

No, it is static and I don't want this value to persist between games

teinarss replied:
This is a bigger problem in the engine structure, that i have mentioned in the past, and causes us to add logic to classes by the life span of them and not where it really belongs, But it would be nice if we dont add more stuff here when i already have spent some time making this class much more cleaner.

PunkPun replied:
I understand that and fully agree. This was just a temporary way to get mutes to work and I will find a better place for it. Having said that it would be best to focus on merging dependancies before moving onto this pr

abcdefg30 replied:
> No, it is static and I don't want this value to persist between games

`IngameChatLogic` would need to clear it in `Disposed` which I assume runs when the game ends?

PunkPun replied:
> > No, it is static and I don't want this value to persist between games
> 
> `IngameChatLogic` would need to clear it in `Disposed` which I assume runs when the game ends?

It runs when the game ends, and when player exist the game. So basically mutes would disappear when you enter after-game chat, what is definitely not desired

PunkPun replied:
I moved `MutedPlayers` to `TextNotificationManager` and added a call to `Dispose()` to `IngameMenuLogic`

dragunoff commented on mods/cnc/chrome/ingame-infostats.yaml:
```diff
@@ -153,3 +168,13 @@ Container@SKIRMISH_STATS:
+							Checkmark: mute
+							Background:
+							TooltipContainer: TOOLTIP_CONTAINER
+							TooltipText: Mute this player
```
This line (and all the others like it) is not needed anymore because the tooltip text is set in the chrome logic.

Smittytron commented:
Needs rebase

PunkPun commented:
Rebased

teinarss commented:
Needs another rebase.

PunkPun commented:
rebased

Mailaender reviewed (dismissed):
Works as promised.

PunkPun commented:
RA glyph-3x was outdated, fixed it

PunkPun commented:
Updated the PR to add button background when hovering mute and kick buttons. Moved the kick button rightways and added an `Actions` section to host them both. This makes way for the kick button to appear inside regular player previews for #20083


https://user-images.githubusercontent.com/37534529/183461785-123f69eb-a738-427b-bebe-f378eece551f.mov

PunkPun committed:
Add per player mutes

PunkPun committed:
Increase the size of ingame-info panel

And add an Actions section for Kick

PunkPun committed:
Fixup faction and score info UI

And increase spectator name length to match regular players

PunkPun commented:
Now using the icons added in https://github.com/OpenRA/ArtSrc/pull/17
I've also used fluent for the mute tooltip

PunkPun committed:
Add chat on/off icon to glyphs

dragunoff reviewed (dismissed):
Code and UI changes LGTM and I didn't spot any regressions when testing :+1:

abcdefg30 commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/GameInfoStatsLogic.cs:
```diff
@@ -51,17 +58,22 @@ public GameInfoStatsLogic(Widget widget, World world, OrderManager orderManager,
 				playerPanel.Bounds.Height += objectiveGroup.Bounds.Height;
 			}
 
+			// orderManager.LocalClient is null when watching a replay
```
This comment would live better at the `OrderManager.LocalClient` property itself as a `summary`.

PunkPun replied:
added a summary

abcdefg30 commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/GameInfoStatsLogic.cs:
```diff
@@ -159,6 +177,12 @@ public GameInfoStatsLogic(Widget widget, World world, OrderManager orderManager,
+					muteCheckbox.IsChecked = () => TextNotificationsManager.MutedPlayers[client.Index];
+					muteCheckbox.OnClick = () => TextNotificationsManager.MutedPlayers[client.Index] ^= true;
+					muteCheckbox.IsVisible = () => !client.IsBot && client.State != Session.ClientState.Disconnected && client.Index != orderManager.LocalClient?.Index;
+					muteCheckbox.GetTooltipText = () => muteCheckbox.IsChecked() ? "Unmute this player" : "Mute this player";
```
This should use the translations I assume.

PunkPun replied:
not the first time I forgot the code is duplicated

abcdefg30 commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameChatLogic.cs:
```diff
@@ -69,6 +69,9 @@ public IngameChatLogic(Widget widget, OrderManager orderManager, World world, Mo
 			var teamMessage = modData.Translation.GetString(Team);
 			var allMessage = modData.Translation.GetString(All);
 
+			if (TextNotificationsManager.MutedPlayers.Count == 0)
```
Is there a case where there will be non-zero players in the dictionary here?

PunkPun replied:
Yes, this is called every time you open the ingame menu and it has to be here as UI is one of the first things that's loaded into the game

abcdefg30 replied:
Ah, right.

teinarss commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameChatLogic.cs:
```diff
@@ -69,6 +69,9 @@ public IngameChatLogic(Widget widget, OrderManager orderManager, World world, Mo
 			var allMessage = modData.Translation.GetString(All);
 
+			if (TextNotificationsManager.MutedPlayers.Count == 0)
+				orderManager.LobbyInfo.Clients.Do((c) => TextNotificationsManager.MutedPlayers.Add(c.Index, false));
```
foreach()

PunkPun replied:
https://github.com/OpenRA/OpenRA/pull/20100#discussion_r948720165

PunkPun replied:
ok

PunkPun commented:
git compare is failing, using regular foreach() now

teinarss commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/9503433e214b38ca5c9834881ecad12fe69cf703...08dff7f86c58d6b488f091c8ca3d9042984f59b6
